concordium-node/src/grpc2.rs

```diff
@@ -983,12 +984,21 @@ pub mod server {
                 };
 
                 let task = tokio::spawn(async move {
-                    router
+                    let result = router
                         .serve_with_shutdown(
                             std::net::SocketAddr::new(listen_addr, listen_port),
                             shutdown_receiver.map(|_| ()),
                         )
-                        .await
+                        .await;
+                    if let Err(ref err) = result {
+                        // Log an error and notify main thread that an error occured.
+                        error!("A runtime error occurred in the GRPC2 server: {}", err);
+                        error_sender.send(()).expect(
+                            "A runtime error occurred in the GRPC2 server, but it was unable to \
+                             notify the owning thread.",
+                        );
+                    }
+                    result
                 });
                 Ok(Some(Self {
                     task,
```

Comment:
I cannot comment on the actual line below in the "shutdown" function, but the output is not the best now.

In particular if the node fails to start because a port is already occupied the output is now

```
2022-11-24T09:18:06.957760033Z: ERROR: A runtime error occurred in the GRPC2 server: transport error
2022-11-24T09:18:06.957812280Z: INFO: Starting the P2P layer
2022-11-24T09:18:06.958250249Z: INFO: Commencing baking
2022-11-24T09:18:06.958310964Z: INFO: Runner: Starting baker thread
2022-11-24T09:18:06.958352063Z: ERROR: Could not stop the GRPC2 server correctly. Forcing shutdown.
```

In particular the last line is misleading.

We should distinguish two kinds of errors here.

Right now we have this channel that is used to signal the main thread to stop. There are two kinds of writers to this channel.

- the signal handler, which indicates a user decided to kill the node.
- A service has failed to start or has crashed. In that case the `shutdown` function should not even be called since the  server is already dead.

One option for handling this is to not have a channel that just sends a (), but instead send either a boolean or more variants.

One variant means the user has requested a shutdown.

Another means that the grpc2 server has been terminated.

Reply:
Hope it looks better now.